statsmodels/regression/linear_model.py

```diff
@@ -750,8 +750,37 @@ def hessian(self, params, scale=None):
             return -self._wexog_xprod / scale
 
         return hess
+    
 
+    def hessian_obs(self, params, scale=None, diag=True):
+        """
+        hessian second derivative of the loglikelihood for each observation.
+
+        Parameters
+        ----------
+        params : ndarray
+            parameter at which Hessian is evaluated
+        scale : None or float
+            If scale is None, then the default scale will be calculated.
+            Default scale is defined by `self.scaletype` and set in fit.
+            If scale is not None, then it is used as a fixed scale.
+        diag : bool
+            Whether to return the diag of the 3d mat
+
+        Returns
+        -------
+        hessian_obs : ndarray, 2d
+            The second derivative of the loglikelihood function evaluated at
+            params for each observation.
+        """
+
+        if diag:
```

Comment:
I don't think we want to actually create the diagonal matrix here.  Just return the diagonal itself (as a vector), then use it as if it were a diagonal matrix.  The full diagonal matrix ix nxn, too big and wasteful to create.